### ISSUE TITLE
Exception running on one's own dataset

### ISSUE THREAD

**Issue by theneao** [mathpluscode/ImgX-DiffSeg#6]
By imitating Amos_ CT wrote the "data_builder. py" and "config" files of its own 2D dataset, and ran the algorithm using 2D methods without modifying the network structure. However, during the training process, the loss was always negative.

The dataset consists of 5 categories of 2D images labeled 0-4, with each image containing only one or two types of data. The labels for each image are not all 0, and loss also indicates this (different categories randomly appear with loss values, not always nan, while background class loss always exists). At first, it was thought that the loss did not pay attention to nan, but it was found that it had been changed to 0 for nan. At the same time, the jax architecture was the first time I came into contact and it was difficult to debug during jit compilation. Therefore, I have been researching for a long time and have no clue. I hereby seek advice.
Here is my "data_builder. py" and "config" file
[my_dataset.zip](https://github.com/mathpluscode/ImgX-DiffSeg/files/11304369/my_dataset.zip)
Perhaps some configuration information in these two files was written incorrectly, because my dataset is not a medical image, it is a 2D semantic segmentation that does not contain z-axis information. However, I have done my best to modify any areas that I believe are different. I also adapted some of the medical image processing libraries you called, mainly due to the lack of a dimension for addition.
Mainly modified`imgx/datasets/preprocess.py`，such as
```
    image_volume = sitk.ReadImage(str(image_path),sitk.sitkInt8)
    label_volume = sitk.ReadImage(str(label_path),sitk.sitkInt8)
```
```
    image_array = sitk.GetArrayFromImage(image_volume)
    if len(image_array.shape) == 2:
        image_array = np.expand_dims(image_array, axis=0)
        image_volume = sitk.GetImageFromArray(image_array)
```
The program can run normally, but as shown in the figure, there is a large amount of nan in the loss. After running, only the background class loss drops normally, while other classes cannot, which is meaningless.
![image](https://user-images.githubusercontent.com/71068599/233854396-4fbbd3c3-9de7-4e99-b782-a5aca2fc4964.png)
This is an example of annotated images, which only includes four categories and is missing one category.
![image](https://user-images.githubusercontent.com/71068599/233854803-fb7ab1ed-24e8-4f21-a87c-e535d92c89d3.png)
If you could spare some time from your busy schedule to take a look, I would greatly appreciate it. The quality and speed of the code you wrote also impressed me, as compared to Pytorch's diffusion model, it can train very quickly.

**Comment by mathpluscode** [comment]
Hi @theneao, thanks for your message. Please see my responsed below, hope it could be helpful for you.

## nan dice

As in [dice_score](https://github.com/mathpluscode/ImgX-DiffSeg/blob/main/imgx/metric/dice.py#L21), the nan may happen when the class does not exist in the image. Have you looked at the [metrics](https://github.com/mathpluscode/ImgX-DiffSeg/blob/main/imgx/exp/eval.py#L97) `f"class_{i}_proportion_{mask_name}"`  to see when you have nan loss, if the proportion equals zero?

The original codebase / network was assuming the existence of all classes, however, as we used `jnp.nanmean` for [dice loss](https://github.com/mathpluscode/ImgX-DiffSeg/blob/main/imgx/loss/dice.py#L15) it should not be a problem.

## Slow training / convergence

In this version of code, the network followed code in [MONAI](https://docs.monai.io/en/stable/networks.html#basicunet) and therefore it is different from the unet [here](https://github.com/hojonathanho/diffusion/blob/1e0dceb3b3495bbe19116a5e1b3596cd0706c543/diffusion_tf/models/unet.py#L114). Note that `hojonathanho`'s unet contains attention mechanism, which can be easily added using [multi head attention](https://dm-haiku.readthedocs.io/en/latest/api.html?highlight=attention#attention).

Otherwise, maybe you can try to use a different set of hyper parameters. Given it's not medical images, you may consider following [SAM model from Meta](https://arxiv.org/abs/2304.02643) for hyper parameters settings.

Moreover, have you compared with different batch sizes? This is often another important hyper parameters to tune.

**Comment by mathpluscode** [comment]
Hi, I am closing this issue. Please feel free to reopen it if there are more questions. Thanks.
